cmd/zfs/zfs_main.c

```diff
@@ -6772,12 +6773,31 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
 	}
 
 	/*
-	 * Skip any dataset whos name doesn't have a prefix matching ga_dataset.
+	 * Check if ga_filesystem is set; used for recursive mounting
 	 */
-	if (state->ga_dataset != NULL && strncmp(zfs_get_name(zhp),
-	    state->ga_dataset, strlen(state->ga_dataset)) != 0) {
-		zfs_close(zhp);
-		return (0);
+	if (state->ga_filesystem != NULL) {
+		/*
+		 * Skip any dataset with the canmount=off property.
+		 */
+		if (zfs_prop_get_int(zhp, ZFS_PROP_CANMOUNT) == ZFS_CANMOUNT_OFF) {
+			zfs_close(zhp);
+			return (0);
+		}
+
```

Comment:
Why is this here? There are checks elsewhere to exclude filesystems with `canmount=off`, so there's no need to duplicate the logic here.

Reply:
Without this check `-r` will explicitly try mounting `canmount=off` which results in an error.
The check you mentioned is in the mount code.

Reply:
The right place for that check is not here, because it's changing the notional behavior of `get_one_filesystem` beyond your intended scope of "descending from this point". Having the function enumerate filesystems that *include* `canmount=off` when no parent is specified but *omit* the same just because a desired parent was specified is counter-intuitive.

The `canmount` check *still* belongs in `share_mount_one`, but it highlights an inconsistency in your definitions of scope-limited `-a` and `-r`. The idea behind `mount -a` is "mount everything that can be mounted", while an explicit mount is supposed to ignore things like `canmount=noauto` because it's been specifically requested. Nevertheless, `canmount=off` is a hard stop that triggers an error.

Presumably, you introduced `-r` to do the equivalent of

```sh
for fs in <children of some ZFS filesystem>; do
    zfs mount $fs
done
```

but what is the right behavior if you do this and one of the filesystems expressly disallows mounting with `canmount=off`? Is it to ignore the error and continue, or throw an error? The former means that `-r` is a sort of weird `-a` that behaves exactly as `-a` except that it ignores `canmount=noauto` and mounts it anyway. And yet using recursion is effectively an "automatic" operation. However, failing is also problematic, because then it leaves some subset of the desired mounts in a mounted state.

I'd submit that the whole notion of `-r` is ill-formed, and you should abandon the new argument and stick to just adding the option for limiting the scope of `-a` to some subtree. If not, you'll need to separate the logic between `explicit` and `recursive` so you can properly ignore `canmount=off` for recursive operations as well.